File: clients/mobile/android/app/src/androidTest/kotlin/com/sharepaste/android/RoundTripTest.kt
package com.sharepaste.android

import androidx.activity.ComponentActivity
import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.compose.ui.test.onNodeWithTag
import androidx.compose.ui.test.performClick
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.sharepaste.android.ui.Receipt
import com.sharepaste.android.ui.SessionPhase
import com.sharepaste.android.ui.TAG_OFFER
import com.sharepaste.android.ui.entryRecallTag
import com.sharepaste.core.AppException
import com.sharepaste.core.SettingsPatch
import kotlinx.coroutines.runBlocking
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

/**
 * The round trip that justifies the client at all.
 *
 * Copy on the laptop, open the phone, see it and Recall it onto the phone's
 * clipboard; Offer from the phone, see it on the laptop. Against the live Relay
 * with a second real device on the other end of the Pairing — a stub on either
 * side would prove nothing about the bytes.
 *
 * Everything is driven through the shipped objects: the app's `SharepasteRepository`,
 * its `SharepasteViewModel`, its `HistoryScreen`, and the same `appActions(model)`
 * bag `MainActivity` builds. The clipboard assertions read `ClipboardManager`
 * itself, because reaching the platform's clipboard is the whole criterion.
 *
 * Each test pairs its own phone into the run's one inviting User, so nothing here
 * depends on the order JUnit picks.
 *
 * **The in-app Recall does not fetch, and that is the point of ADR 0010.** The
 * test that used to sit here proved the opposite of `RECALL LATEST`: the phone
 * was put down, the other device put an Entry on the Relay that this phone's
 * cache had never held, and the verb handed it over. That verb is now the
 * notification's alone. `RECALL FIRST` hands over the first row of the
 * displayed list, which is a row on screen by construction, so the fetch it
 * lost cannot swap in an Entry the person never saw — and
 * `StandingActionsOnAClosedPhoneTest` still holds the round trip and the
 * fallback that may never be silent, on the surface that kept them.
 */
@RunWith(AndroidJUnit4::class)
class RoundTripTest {

    @get:Rule
    val compose = createAndroidComposeRule<ComponentActivity>()

    private lateinit var phone: PhoneUnderTest
    private lateinit var other: Inviter

    @Before
    fun open() {
        other = Inviter.shared()
        phone = PhoneUnderTest.open(compose, DATABASE)
        // Paired but not yet in the foreground: several of these tests turn on
        // what a phone has and has not seen before its first `onStart`.
        phone.pairWithCode(other, "round-trip test phone")
    }

    @After
    fun close() = phone.close()

    /**
     * Something copied on the other device reaches the phone on a resume, and
     * Recalling it puts it on this phone's clipboard.
     *
     * The Entry is put on the Relay while the phone is closed, which is the shape
     * of the real thing: sync is foreground-only, so the first `onStart` is what
     * brings it in.
     */
    @Test
    fun an_entry_from_another_device_arrives_on_resume_and_recalls_onto_this_clipboard() {
        val copied = "https://example.invalid/from-the-laptop-${System.currentTimeMillis()}"
        other.offerAndWaitForUpload(copied)
        Evidence.log("other device  = offered and uploaded an Entry while this phone was closed")

        phone.enterForeground()
        phone.await("in contact after onStart") { it.session is SessionPhase.InContact }
        val arrived = phone.awaitEntry("the other device's Entry must be backfilled") {
            it.preview == copied
        }
        Evidence.log("backfilled    = id=${arrived.id} preview=${arrived.preview}")
        assertFalse("a decryptable Entry must not be marked", arrived.undecryptable)
        assertEquals(
            "this phone's own Device id has to be known, or every row claims an Origin",
            runBlocking { phone.repo.listPairings() }.single { it.userId == phone.userId }.deviceId,
            phone.state.ownDeviceId,
        )
        assertTrue(
            "this Entry came from the other device, so it has an Origin to show",
            arrived.deviceId != phone.state.ownDeviceId,
        )

        // Overwritten first, so a pass cannot be an accident of whatever the
        // clipboard already happened to hold.
        phone.clip.putText("something else entirely")
        phone.scrollTo(entryRecallTag(arrived.id))
        compose.onNodeWithTag(entryRecallTag(arrived.id)).performClick()
        val recalled = phone.awaitReceipt("the Recall must report itself") {
            it is Receipt.Recalled
        } as Receipt.Recalled
        assertEquals(
            "the Recall must say which Entry it handed over, and the Preview is how it says it",
            copied,
            recalled.preview,
        )

        val onClipboard = phone.clip.requireText("after Recalling the backfilled Entry")
        Evidence.log("recalled      = clipboard now holds ${onClipboard.length} chars")
        assertEquals("the Recall must put that Entry's plaintext on the clipboard", copied, onClipboard)
    }

    /**
     * Something Offered on the phone reaches the Relay and the other device can
     * read it.
     *
     * Read back with the other device's own `recall_latest`, which always performs
     * the round trip — so the assertion is about what the Relay holds now, not
     * about anything either device had cached.
     */
    @Test
    fun an_entry_offered_here_is_readable_by_the_other_device() {
        phone.enterForeground()
        phone.await("in contact before offering") { it.session is SessionPhase.InContact }

        val offered = "offered-from-the-phone-${System.currentTimeMillis()}"
        phone.clip.putText(offered)
        compose.onNodeWithTag(TAG_OFFER).performClick()
        val taken = phone.awaitReceipt("the Offer must be taken") { it is Receipt.Offered }
        Evidence.log("offered       = $taken")

        awaitRelayNewest("the phone's Offer must reach the Relay", offered)
        Evidence.log("other device  = read it back off the Relay: ${offered.take(24)}…")
    }

    /**
     * An Offer is honoured with capture disabled.
     *
     * `capture_enabled` governs Watched Capture, which a phone never performs.
     * Every Entry a phone produces is an Offered Capture — the person handed the
     * content over — and refusing that because of a desktop's watcher setting
     * would be indefensible.
     */
    @Test
    fun an_offer_is_honoured_with_capture_disabled() {
        // Only `captureEnabled` is patched. `autostart`, `hotkey` and
        // `updateCheckEnabled` are desktop concerns carried on the same row, and
        // a phone that named them would clear them.
        val settings = runBlocking { phone.repo.updateSettings(SettingsPatch(captureEnabled = false)) }
        assertFalse("capture must really be off for this to prove anything", settings.captureEnabled)

        phone.enterForeground()
        phone.await("in contact") { it.session is SessionPhase.InContact }

        val offered = "offered-with-capture-disabled-${System.currentTimeMillis()}"
        phone.clip.putText(offered)
        compose.onNodeWithTag(TAG_OFFER).performClick()
        val taken = phone.awaitReceipt("the Offer must be taken with capture disabled") {
            it is Receipt.Offered
        }
        Evidence.log("capture off   = $taken (capture_enabled=false)")

        awaitRelayNewest("an Offer made with capture disabled must still reach the Relay", offered)
    }

    /**
     * ADR 0011, across the two devices that make it a decision at all: a Recall
     * of a buried Entry puts it at the head of the History here **and** there.
     *
     * The other device is asked with `recall_latest`, which always performs the
     * round trip, so what it hands back is what the Relay considers the head
     * *now* rather than anything it had cached. Ordering that agrees across
     * devices is the whole point; a phone that reordered only its own list
     * would pass a weaker version of this test and ship the wrong feature.
     *
     * Nothing about the Entry itself moves — same id, same plaintext — which is
     * what "nothing is created and nothing is duplicated" means in practice.
     */
    @Test
    fun recalling_a_buried_entry_puts_it_at_the_head_on_both_devices() {
        val stamp = System.currentTimeMillis()
        val buried = "buried-entry-$stamp"
        val newer = "captured-since-$stamp"
        val buriedId = other.offerAndWaitForUpload(buried)
        other.offerAndWaitForUpload(newer)

        phone.enterForeground()
        phone.await("in contact") { it.session is SessionPhase.InContact }
        val onScreen = phone.awaitEntry("the buried Entry must be cached") { it.preview == buried }
        phone.awaitEntry("and the one captured after it") { it.preview == newer }
        assertEquals("the phone holds it under the Relay's own id", buriedId, onScreen.id)
        assertEquals(
            "precondition: the Entry captured last leads the History until something is used",
            newer,
            runBlocking { phone.repo.listHistory(phone.userId!!) }.first().preview,
        )
        Evidence.log("before recall = head is $newer, $buried is buried below it")

        phone.scrollTo(entryRecallTag(onScreen.id))
        compose.onNodeWithTag(entryRecallTag(onScreen.id)).performClick()
        phone.awaitReceipt("the Recall must report itself") { it is Receipt.Recalled }

        awaitHistoryHead("the recalled Entry must lead this phone's History", buried)
        Evidence.log("after recall  = this phone's head is $buried")
        awaitRelayNewest("and the Relay's, so every other device reorders with it", buried)
        Evidence.log("other device  = the Relay hands back $buried as its head too")

        // Counted rather than sized: the run's one inviting User is shared, so
        // another test's Entries may sit in this History too. What a use must
        // never do is *add* one.
        val after = runBlocking { phone.repo.listHistory(phone.userId!!) }
        assertEquals(
            "a use creates nothing, so the recalled text appears once and not twice",
            1,
            after.count { it.preview == buried },
        )
        assertEquals("and the Entry at the head is the one that was always there", buriedId, after.first().id)
    }

    /** Poll this phone's own History until [expected] is its head. */
    private fun awaitHistoryHead(what: String, expected: String) {
        val deadline = System.nanoTime() + PhoneUnderTest.TIMEOUT_SECONDS * 1_000_000_000L
        while (System.nanoTime() < deadline) {
            val head = runBlocking { phone.repo.listHistory(phone.userId!!) }.firstOrNull()?.preview
            if (head == expected) return
            Thread.sleep(250)
        }
        throw AssertionError("$what: it never reached the head")
    }

    /** Poll the other device until the Relay's newest Entry is [expected]. */
    private fun awaitRelayNewest(what: String, expected: String) {
        val deadline = System.nanoTime() + PhoneUnderTest.TIMEOUT_SECONDS * 1_000_000_000L
        while (System.nanoTime() < deadline) {
            val newest = try {
                other.newestOnRelay()
            } catch (e: AppException) {
                null
            }
            if (newest == expected) return
            Thread.sleep(250)
        }
        throw AssertionError("$what: the other device never saw it")
    }

    private companion object {
        const val DATABASE = "round-trip-proof.db"
    }
}

File: clients/mobile/ios/Tests/SharepasteFacadeTests/RoundTripTest.swift
import Foundation
import SharepasteCore
import SharepasteKit
import XCTest

/// The round trip that justifies the client at all.
///
/// Copy on the laptop, open the phone, see it and Recall it onto the phone's
/// pasteboard; Offer from the phone, see it on the laptop. Against the live
/// relay with a second real device on the other end of the Pairing — a stub on
/// either side would prove nothing about the bytes.
///
/// Everything is driven through the shipped ``SharepasteRepository``, and the
/// pasteboard assertions read the ``TestPasteboard`` the phone handed it.
/// `UIPasteboard` is not reachable from a host-less test bundle at all; the
/// reasoning, and the hang that established it, are on that type. What is under
/// test is unchanged by the substitution: the Clipboard trait is a foreign
/// trait precisely so that the shell decides what a pasteboard is, and the
/// crossing being asserted is the core reaching a Swift object.
///
/// **A Recall's write is the core's, not the shell's.** Nothing in
/// ``SharepasteRepository`` writes the pasteboard on a Recall —
/// `recall_latest` calls `write_clipboard` inside the core
/// (`clients/core/src/facade.rs`), so the assertion below that the plaintext
/// arrived at the Clipboard trait is evidence about the *core's* behaviour.
/// That matters beyond this test: ticket 07 says neither App Intent touches the
/// pasteboard, and on this path one of them does.
///
/// **What does not port.** On Android these tests now assert a Receipt and the
/// Preview it names. A Receipt is a Toast — it is UI, and spec row 10 buys none
/// — so what is asserted here is the facade half beneath it: that the Recall
/// handed back the Entry only the relay had, that it was not served from the
/// cache, and that the plaintext reached the pasteboard. The Preview the Receipt
/// would have shown is checked too, through
/// ``SharepasteRepository/previewOf(userId:entryId:)``, because that method
/// exists for the Receipt and would otherwise have no test at all.
///
/// Each test pairs its own phone into the run's one inviting device, over a
/// database of its own, so nothing here depends on the order XCTest picks.
final class RoundTripTest: XCTestCase {

    private var phone: PhoneUnderTest!

    override func tearDown() async throws {
        await phone?.close()
        phone = nil
    }

    /// Something copied on the other device reaches the phone on a resume, and
    /// Recalling it puts it on this phone's pasteboard.
    ///
    /// The Entry is put on the relay while the phone is closed, which is the
    /// shape of the real thing: sync is foreground-only, so the first resume is
    /// what brings it in.
    func testAnEntryFromAnotherDeviceArrivesOnResumeAndRecallsOntoThisPasteboard() async throws {
        let other = try await Inviter.shared()
        let userId = try await openPhone("backfill-proof.db", label: "round-trip test phone")

        let copied = "https://example.invalid/from-the-laptop-\(Self.stamp)"
        try await other.offerAndWaitForUpload(copied)

        _ = try await phone.enterForeground()
        try await phone.awaitInContact(userId)
        let arrived = try await phone.awaitEntry(
            "the other device's Entry must be backfilled",
            userId: userId,
            previewing: copied
        )
        XCTAssertFalse(arrived.undecryptable, "a decryptable Entry must not be marked")

        let ownDeviceId = try await phone.repo.listPairings()
            .first { $0.userId == userId }?.deviceId
        XCTAssertNotNil(
            ownDeviceId,
            "this phone's own Device id has to be known, or every row claims an Origin"
        )
        XCTAssertNotEqual(
            arrived.deviceId,
            ownDeviceId,
            "this Entry came from the other device, so it has an Origin to show"
        )

        // Overwritten first, so a pass cannot be an accident of whatever the
        // pasteboard already happened to hold.
        phone.pasteboard.put("something else entirely")
        try await phone.repo.recall(userId: userId, entryId: arrived.id)

        XCTAssertEqual(
            phone.pasteboard.written,
            [copied],
            "the Recall must hand that Entry's plaintext to the pasteboard, exactly once"
        )
        let preview = await phone.repo.previewOf(userId: userId, entryId: arrived.id)
        XCTAssertEqual(
            preview,
            copied,
            "the Receipt names the Entry by its Preview, and this is the read behind it"
        )
    }

    /// Something Offered on the phone reaches the relay and the other device can
    /// read it.
    ///
    /// Read back with the other device's own `recallLatest`, which always
    /// performs the round trip — so the assertion is about what the relay holds
    /// now, not about anything either device had cached.
    func testAnEntryOfferedHereIsReadableByTheOtherDevice() async throws {
        let other = try await Inviter.shared()
        let userId = try await openPhone("offer-proof.db", label: "offer test phone")

        _ = try await phone.enterForeground()
        try await phone.awaitInContact(userId)

        let offered = "offered-from-the-phone-\(Self.stamp)"
        phone.pasteboard.put(offered)
        let attempt = try await phone.repo.offerPasteboard()
        guard case .settled(let offeredTo, let outcome) = attempt else {
            return XCTFail("a paired phone must not report itself unpaired: \(attempt)")
        }
        XCTAssertEqual(offeredTo, userId, "an Offer goes to the Active Pairing")
        guard case .queued = outcome else {
            return XCTFail("the Offer must be taken, and this one was \(outcome)")
        }

        let seen = await other.awaitNewestOnRelay(offered)
        XCTAssertTrue(seen, "the phone's Offer never reached the relay")
    }

    /// **Recall Latest fetches. It does not read the cache.**
    ///
    /// Proven the only way it can be: the phone is put down, the other device
    /// puts an Entry on the relay that this phone's cache has *never* held, and
    /// Recall Latest is asked with no session running. If it read the cache it
    /// would hand over the older Entry; it hands over the new one.
    ///
    /// The second half is `fromCache`, which is the fact ADR 0007 says may never
    /// be silent. A round trip that succeeded must report `false` — an
    /// authoritative answer reported as stale is a phone crying wolf about the
    /// one thing it has to be believed on.
    func testRecallLatestFetchesAnEntryThisPhoneHasNeverCached() async throws {
        let other = try await Inviter.shared()
        let userId = try await openPhone("recall-latest-proof.db", label: "recall latest phone")

        let older = "older-entry-\(Self.stamp)"
        try await other.offerAndWaitForUpload(older)
        _ = try await phone.enterForeground()
        try await phone.awaitInContact(userId)
        _ = try await phone.awaitEntry(
            "the older Entry must be cached first",
            userId: userId,
            previewing: older
        )

        // Put the phone down. Nothing syncs now, which is the entire sync model.
        try await phone.leaveForeground()

        let newer = "newer-entry-the-cache-has-never-seen-\(Self.stamp)"
        try await other.offerAndWaitForUpload(newer)
        try await Task.sleep(nanoseconds: Self.settle)

        let cached = try await phone.repo.listHistory(userId: userId)
        XCTAssertFalse(
            cached.contains { $0.preview == newer },
            "the newer Entry reached this phone's cache; with a live session this test would "
                + "prove nothing about fetching"
        )
        XCTAssertEqual(cached.first?.preview, older, "a cache read would hand over this one")

        phone.pasteboard.put("something else entirely")
        let attempt = try await phone.repo.recallLatestOnActivePairing()
        guard case .done(let recalledFor, let entryId, _, let fromCache) = attempt else {
            return XCTFail("a paired phone must not report itself unpaired: \(attempt)")
        }
        XCTAssertEqual(recalledFor, userId)
        XCTAssertFalse(
            fromCache,
            "the relay was reachable, so this answer is authoritative and must not be reported "
                + "as stale"
        )
        XCTAssertEqual(
            phone.pasteboard.written.last,
            newer,
            "Recall Latest read the cache instead of fetching"
        )
        let preview = await phone.repo.previewOf(userId: userId, entryId: entryId)
        XCTAssertEqual(
            preview,
            newer,
            "the Receipt must name the Entry only the relay had; a cache read would have named "
                + "the older one"
        )
    }

    /// An Offer is honoured with capture disabled.
    ///
    /// `captureEnabled` governs Watched Capture, which a phone never performs.
    /// Every Entry a phone produces is an Offered Capture — the person handed
    /// the content over — and refusing that because of a desktop's watcher
    /// setting would be indefensible.
    ///
    /// The setting is written **before** the phone exists, through a facade of
    /// this test's own that is released before the repository opens. The
    /// chokepoint exposes no `updateSettings` — nothing on this phone's screens
    /// can change a desktop's watcher, which is the right surface — and two
    /// facades open on one SQLite file is not something to arrange on purpose.
    func testAnOfferIsHonouredWithCaptureDisabled() async throws {
        let database = "capture-disabled-proof.db"
        let directory = try freshDatabase(named: database)
        try autoreleasepool {
            let core = try Sharepaste.open(
                dbPath: directory.appendingPathComponent(database).path,
                keychain: InMemoryKeychain(),
                clipboard: NoClipboard(),
                events: SilentSink(),
                requireHttps: false
            )
            // Only `captureEnabled` is patched. `autostart`, `hotkey` and
            // `updateCheckEnabled` are desktop concerns carried on the same row,
            // and a phone that named them would clear them.
            let settings = try core.updateSettings(patch: SettingsPatch(captureEnabled: false))
            XCTAssertFalse(
                settings.captureEnabled,
                "capture must really be off for this to prove anything"
            )
        }

        let other = try await Inviter.shared()
        phone = try PhoneUnderTest.open(databaseName: database, fresh: false)
        let userId = try await phone.pair(with: other, label: "capture disabled phone")

        _ = try await phone.enterForeground()
        try await phone.awaitInContact(userId)

        let offered = "offered-with-capture-disabled-\(Self.stamp)"
        let attempt = try await phone.repo.offerText(offered)
        guard case .settled(_, let outcome) = attempt, case .queued = outcome else {
            return XCTFail("the Offer must be taken with capture disabled: \(attempt)")
        }

        let seen = await other.awaitNewestOnRelay(offered)
        XCTAssertTrue(seen, "an Offer made with capture disabled must still reach the relay")
    }

    /// A phone of this test's own, paired into the run's inviting device and
    /// left resting.
    private func openPhone(_ database: String, label: String) async throws -> String {
        let inviter = try await Inviter.shared()
        phone = try PhoneUnderTest.open(databaseName: database)
        return try await phone.pair(with: inviter, label: label)
    }

    /// How long the phone is left closed while the other device offers.
    ///
    /// The Entry is on the relay before the wait starts, so this is only the
    /// window in which a session that had not come down would have brought it
    /// in.
    private static let settle: UInt64 = 3_000_000_000

    /// Distinct content per test, so one run's Entries cannot be mistaken for
    /// another's on a relay this suite shares with itself.
    private static var stamp: Int {
        Int(Date().timeIntervalSince1970 * 1000)
    }
}
